oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/banyandb/client/impl/BanyanDBGrpcClient.java

```diff
@@ -0,0 +1,173 @@
+            BanyanDBEntity.BanyanDBEntityBuilder entityBuilder = BanyanDBEntity.builder();
+            for (final Query.TypedPair typedPair : entity.getFieldsList()) {
+                if (typedPair.hasIntPair()) {
+                    final Query.IntPair pair = typedPair.getIntPair();
+                    entityBuilder.field(pair.getKey(), pair.getValuesList().get(0));
+                } else if (typedPair.hasStrPair()) {
+                    final Query.StrPair pair = typedPair.getStrPair();
+                    entityBuilder.field(pair.getKey(), pair.getValuesList().get(0));
+                }
+            }
+            entityBuilder.entityId(entity.getEntityId());
+            entityBuilder.binaryData(entity.getDataBinary().toByteArray());
+            entityBuilder.timestampSeconds(entity.getTimestamp().getSeconds());
+            entityBuilder.timestampNanoSeconds(entity.getTimestamp().getNanos());
+            r.getEntities().add(entityBuilder.build());
+        }
+
+        return r;
+    }
+
+    @Override
+    public void writeEntity(List<TraceWriteRequest> data) {
+        StreamObserver<org.apache.skywalking.banyandb.Write.WriteRequest> observer = this.asyncStub.withDeadlineAfter(30, TimeUnit.SECONDS)
+                .write(new StreamObserver<Write.WriteResponse>() {
+                    @Override
+                    public void onNext(Write.WriteResponse writeResponse) {
+                    }
+
+                    @Override
+                    public void onError(Throwable throwable) {
+                        log.error("fail to send request", throwable);
+                    }
+
+                    @Override
+                    public void onCompleted() {
+                    }
+                });
+
+        for (final TraceWriteRequest entity : data) {
+            try {
+                Write.EntityValue entityValue = Write.EntityValue.newBuilder()
+                        .addAllFields(buildWriteFields(entity.getFields()))
+                        .setDataBinary(ByteString.copyFrom(entity.getDataBinary()))
+                        .setTimestamp(Timestamp.newBuilder().setSeconds(entity.getTimestampSeconds()).setNanos(entity.getTimestampNanos()).build())
+                        .setEntityId(entity.getEntityId()).build();
+                observer.onNext(Write.WriteRequest.newBuilder()
+                        .setMetadata(Database.Metadata.newBuilder().setName(entity.getName()).setGroup(entity.getGroup()).build())
+                        .setEntity(entityValue).build());
+            } catch (Throwable t) {
+                log.error("fail to convert entityValue and send data");
+            }
+        }
+
+        observer.onCompleted();
+    }
+
+    @VisibleForTesting
+    public static List<Write.Field> buildWriteFields(List<Object> fieldList) {
+        return fieldList.stream().map(BanyanDBGrpcClient::buildField).collect(Collectors.toList());
+    }
+
+    static Write.Field buildField(Object value) {
```

Comment:
Don't use `instance of`, use overload for different types of parameters, which have much better performance and easier to read.

Reply:
No `instanceof` now.